src/upload/index.ts

```diff
@@ -0,0 +1,201 @@
+import {
+  HeadObjectCommand,
+  NotFound,
+  PutObjectCommand,
+  S3Client,
+} from "@aws-sdk/client-s3";
+import * as zlib from "zlib";
+import fs from "fs";
+
+import { LightBlockCache, lightBlockCache } from "@/cache";
+import { LightBlock } from "@/models/lightstreamer";
+import { logger } from "@/utils/logger";
+
+class UploadError extends Error {}
+
+export type BlockRange = {
+  start: number;
+  end: number;
+};
+
+export type BlockFile = {
+  file: string;
+  manifest: string;
+};
+
+export class LightBlockUpload {
+  private cache: LightBlockCache;
+  private s3Client: S3Client;
+  private chunkSizeMb: number;
+  private bucket: string;
+
+  constructor(cache: LightBlockCache) {
+    this.cache = cache;
+    if (!process.env["BUCKET_ENDPOINT"]) {
+      throw new UploadError("BUCKET_ENDPOINT not set");
+    }
+    if (!process.env["BUCKET_ACCESS_KEY_ID"]) {
+      throw new UploadError("BUCKET_ACCESS_KEY_ID not set");
+    }
+    if (!process.env["BUCKET_SECRET_ACCESS_KEY"]) {
+      throw new UploadError("BUCKET_SECRET_ACCESS_KEY not set");
+    }
+    if (!process.env["BUCKET_NAME"]) {
+      throw new UploadError("BUCKET_NAME not set");
+    }
+    if (!process.env["UPLOAD_CHUNK_SIZE_MB"]) {
+      throw new UploadError("UPLOAD_CHUNK_SIZE_MB not set");
+    }
+    this.chunkSizeMb = parseInt(
+      process.env["UPLOAD_CHUNK_SIZE_MB"] as string,
+      10,
+    );
+    this.bucket = process.env["BUCKET_NAME"];
+    this.s3Client = new S3Client({
+      region: "auto",
+      endpoint: process.env["BUCKET_ENDPOINT"],
+      credentials: {
+        accessKeyId: process.env["BUCKET_ACCESS_KEY_ID"],
+        secretAccessKey: process.env["BUCKET_SECRET_ACCESS_KEY"],
+      },
+    });
+  }
+
+  async upload(): Promise<void> {
+    // eslint-disable-next-line no-constant-condition
+    const fileName = "blocks";
+    const currentUploadSize = (await this.getFileSize(fileName)) || 0;
+    logger.info(
+      `Current file uploaded size: ${currentUploadSize}, creating new upload...`,
+    );
+    const files = await this.createBlockFiles(fileName, currentUploadSize);
+
+    logger.info(`Upload: begin...`);
+    const gzip = await this.gzipFile(files.file, `${files.file}.gz`);
+    await this.uploadFile(gzip, "application/gzip");
+    logger.info(`Upload: gzip file complete`);
+
+    await this.uploadFile(files.file, "application/octet-stream");
+    logger.info(`Upload: binary file complete`);
+
+    const gzipManifest = await this.gzipFile(
+      files.manifest,
+      `${files.manifest}.gz`,
+    );
+    await this.uploadFile(gzipManifest, "application/gzip");
+    logger.info(`Upload: manifest file complete`);
+
+    await this.upload();
```

Comment:
Is the intent here to upload the `blocks` file in a continuous fashion? If so, I have some questions:

- Isn't this kind of recursion going to result in a huge stack of nested promises? Perhaps `return this.upload();` may be an easy way to fix the problem. Another one would be to use a `while (true)` loop.

- Should we put a delay in between calls?

- If an error occurs anywhere in this method, it looks like the loop will be broken and never restarted. Is it the case, or is there some logic elsewhere to recover from failures?

Reply:
- I can flip to use `return this.upload()`, this is a good catch and an oversight on my part relative to async recursion.
- No delay is needed, as we can start to assemble the next upload immediately. The line[ here](https://github.com/iron-fish/wallet-server/pull/54/files/505116de9e9adbc321bbf43aa4caaaeefcd68c19#r1540352080) will only occur after N-MB of blocks are added.
- There is no retry logic,  how would you suggest handling, on failure just retry with backoff?

Reply:
> There is no retry logic, how would you suggest handling, on failure just retry with backoff?

The simplest thing would probably to just catch & log the exception, and then re-run `upload()` (which will generate a new `blocks` file). However one problem with this is that if R2 is highly unreliable, we risk not being able to upload most of the `blocks` files generated.

What I would do is use a simple approach and add metrics that we can monitor to see what the failure rate is. If we find that the failure rate is causing disruption, we can adjust the strategy.

Reply:
I will create a datadog monitor that will watch the `latest.json` file (which includes the latest file paths) and verify the `timestamp` field in it is newer than now minus N minutes.